### PR TITLE
feat: support sqlite open v2

### LINKED ISSUE #131: How to open database with immutable flag?
Hi,

this is more question than bug report.
I need to read firefox's bookmark database. But when FF is started the database is locked.

```
➜  hbo9xv4w.default sqlite3 places.sqlite
SQLite version 3.37.0 2021-11-27 14:13:22
Enter ".help" for usage hints.
sqlite> .tables
Error: database is locked
sqlite> 
```

When I open  database with immutable flag, I can read items from database:
```
➜  hbo9xv4w.default sqlite3 'file:places.sqlite?immutable=1'
SQLite version 3.37.0 2021-11-27 14:13:22
Enter ".help" for usage hints.
sqlite> .tables
moz_anno_attributes                    
moz_annos                              
moz_bookmarks 
...
```

Is it possible to open database using `sqlite.lua` library in immutable manner? 

Thank you.

### PR DESCRIPTION
**Pull request by kkharji**
### Purpose

Enable users to control how should the sqlite database opened via opts.open_mode, while enabling support for uri processing https://sqlite.org/c3ref/open.html#urifilenamesinsqlite3open

- [ ] Fix docgen
- [x] Fix tests of fields types

Fix #131

### CODE REVIEW

File: lua/sqlite/defs.lua [hunk 1]
```diff
@@ -680,6 +695,8 @@ M.connect = function(uri, opts)
   return conn[0]
 end
 
+M.connect([[file:/Users/tami/firefox_places.sqlite]], { open_mode = "r" })
```
Comment (Conni2461) [inline lua/sqlite/defs.lua:698]: shouldnt be here

File: lua/sqlite/defs.lua [hunk 2]
```diff
@@ -653,18 +653,33 @@ M.last_errcode = function(conn_ptr)
...
+  local open_mode = opts.open_mode
+
+  if type(open_mode) == "table" then
+    open_mode = bit.bor(unpack(open_mode))
+  else
+    open_mode = M.open_modes[open_mode or "rwc"]
+  end
+
+  local code = clib.sqlite3_open_v2(uri, conn, open_mode, nil)
+  opts.open_mode = nil
```
Comment (Conni2461) [inline lua/sqlite/defs.lua:679]: why do you nil it?

Comment (tami5) [reply lua/sqlite/defs.lua:679]: because of the next processing of opts, to execute `pragma`

**Comment by tami5** [comment]
okay that just weird. on my machine field types comes out uppercased but here lowercased.

**Comment by dhruvmanila** [comment]
Hey @tami5, is there anything blocking this? Do you need any help? 😄

**Comment by tami5** [comment]
Hey @dhruvmanila, thanks for asking, well some tests started occasionally failing. unpredictable behavior of sort, see https://github.com/tami5/sqlite.lua/runs/4653063219?check_suite_focus=true. So I'm kinda worried that merging this would cause unwanted side effect. 

We can be lazy and just check if `opts.open_mode` is provide and thus use open_v2. This would make the PR mergable without fear of some unwanted side effects .

Other than that the issue with docgen. 

I'm quite busy at the moment feel free to take over, thank you

**Comment by dhruvmanila** [comment]
Alright, I might look into it during the weekend, can't guarantee any fixes 😅

**Comment by bliiir** [comment]
Hi guys. I'd like to express my interest in this getting merged as I would enble the use of @dhruvmanila's telescope-bookmarks plugin with Firefox. Thanks for all the hard work you have done and let me know if there is anything I can do to help

**Comment by kkharji** [comment]
@bliiir the pr was blocked for already resolved issue, I just totally forgot about it. Thanks for gesture, I appreciate it

**Comment by bliiir** [comment]
Awesome. Works great - thanks a lot!

**Pull request merged**